streamflow-api.yaml

```diff
@@ -0,0 +1,768 @@
+
+    StreamflowOrderState:
+      description: State for a streamflow order
+      allOf:
+        - type: object
+          properties:
+            id:
+              type: integer
+              example: 5
+            status:
+              type: string
+              enum: [pending, cancelled, fulfilled, deprecated]
+              example: fulfilled
+        - $ref: '#/components/schemas/StreamflowOrderStateNew'
+
+    StreamflowOrderHypermedia:
+      description: Hypermedia for a streamflow order
+      type: object
+      properties:
+        links:
+          allOf:
+            - $ref: '#/components/schemas/HypermediaLinkList'
+            - example:
+                - rel: 'self'
+                  uri: 'http://apis.paccifcclimate.org/streamflow/orders/5'
+                - rel: 'cancel'
+                  uri: 'http://apis.paccifcclimate.org/streamflow/orders/5'
+                - rel: 'hydromodel_output'
+                  uri: 'http://apis.paccifcclimate.org/hydromodel_outputs/4'
+                - rel: 'results/metadata'
+                  uri: 'http://apis.paccifcclimate.org/streamflow/results/6'
+                - rel: 'results/annual-max'
+                  uri: 'http://apis.paccifcclimate.org/streamflow/results/6/annual-max'
+                - rel: 'results/annual-mean'
+                  uri: 'http://apis.paccifcclimate.org/streamflow/results/6/annual-mean'
+                - rel: 'results/annual-cycle'
+                  uri: 'http://apis.paccifcclimate.org/streamflow/results/6/annual-cycle'
+
+    StreamflowOrder:
+      description: Representation of a streamflow order - state + hypermedia.
+      allOf:
+        - $ref: '#/components/schemas/StreamflowOrderState'
+        - $ref: '#/components/schemas/StreamflowOrderHypermedia'
+
+    # List of streamflow orders
+
+    StreamflowOrderListItem:
+      description: List of streamflow results
+      type: object
+      allOf:
+        - type: object
+          properties:
+            uri:
+              type: string
+              description: URI of streamflow order
+              example: 'http://apis.paccifcclimate.org/streamflow/orders/5'
+        - $ref: '#/components/schemas/StreamflowOrderState'
+
+    StreamflowOrderList:
+      description: List of streamflow results
+      type: array
+      items:
+        $ref: '#/components/schemas/StreamflowOrderListItem'
+      example:
+        - uri: 'http://apis.paccifcclimate.org/streamflow/orders/5'
+          hydromodel_output_id: 4
+          longitude: -123.5
+          latitude: 48.5
+          notification:
+            method: email
+            address: abc@example.ca
+        - uri: 'http://apis.paccifcclimate.org/streamflow/orders/6'
+          hydromodel_output_id: 5
+          longitude: -120.0
+          latitude: 50.2
+          notification:
+            method: email
+            address: def@example.ca
+
+    ### Streamflow results
+
+    # Individual streamflow results
+
+    StreamflowResultState:
```

Comment:
Temporal coverage information would be useful. A start date, a stop date, and a resolution, maybe? Temporal coverage info would mean you could actually figure out what dates to request for the annual cycle resource.

Reply:
Would that temporal coverage also (or alternatively) be part of the hydromodel_output, which I believe determines the temporal coverage of the computed result?

(There's no reason we can't add temporal metadata to both. I tend to want to not duplicate information, but this isn't a database that should be normalized, but information resources that are meant to be useful!)

Reply:
Also, what do you recommend as the resolution datum? 
- A string enum like daily, monthly, yearly?
- An exact interval specified either in seconds or by an ISO8601 duration string?

Reply:
I think putting temporal coverage in either the result metadata or the hydromodel output metadata would be fine, as long as it's accessible somewhere. It is theoretically possible to generate a result with a shorter timespan than the hydromodel output, but I can't imagine why we'd want to, and don't see any reason to build in support for that possibility out of the gate.

------

Upon further reflection, I think I was wrong. The result metadata does not need a time resolution attribute because I can't see any reason a client would need that information. A client never works with the raw result data and doesn't need details of its representation in the datafile; the client only interacts with the data in the form of the aggregated annual-mean, annual-max, and annual-cycle endpoints.